cmd/zhack.c

```diff
@@ -485,132 +485,312 @@ zhack_do_feature(int argc, char **argv)
+    uint64_t abdsize)
+{
+	zio_cksum_t verifier;
+	zio_cksum_t actual_cksum;
+	zio_cksum_t expected_cksum;
+	zio_checksum_info_t *ci;
+	abd_t *abd;
+	ssize_t err;
+
+	ZIO_SET_CHECKSUM(&verifier, offset, 0, 0, 0);
+
+	if (byteswap) {
+		byteswap_uint64_array(&verifier, sizeof (zio_cksum_t));
+		eck->zec_magic = BSWAP_64(eck->zec_magic);
+	}
+
+	expected_cksum = eck->zec_cksum;
+	eck->zec_cksum = verifier;
+
+	ci = &zio_checksum_table[ZIO_CHECKSUM_LABEL];
+	abd = abd_get_from_buf(data, abdsize);
+	ci->ci_func[byteswap](abd, abdsize, NULL, &actual_cksum);
+	abd_free(abd);
+
+	if (byteswap)
+		byteswap_uint64_array(&expected_cksum, sizeof (zio_cksum_t));
+
+	if (ZIO_CHECKSUM_EQUAL(actual_cksum, expected_cksum))
+		return (B_FALSE);
+
+	eck->zec_cksum = actual_cksum;
+
+	err = pwrite64(fd, data, abdsize, offset);
+	if (err == -1) {
+		(void) fprintf(stderr, "error: cannot write label %d: %s\n",
+		    l, strerror(errno));
+		return (B_FALSE);
+	} else if (err != abdsize) {
+		(void) fprintf(stderr, "error: bad write size label %d\n", l);
+		return (B_FALSE);
+	} else {
+		(void) fprintf(stderr,
+		    "label %d: wrote %" PRIu64 " bytes at offset %" PRIu64 "\n",
+		    l, abdsize, offset);
+	}
+
+	return (B_TRUE);
+}
+
+#define	ASHIFT_UBERBLOCK_SHIFT(ashift)	\
+	MIN(MAX(ashift, UBERBLOCK_SHIFT), \
+	MAX_UBERBLOCK_SHIFT)
+#define	ASHIFT_UBERBLOCK_SIZE(ashift) \
+	(1ULL << ASHIFT_UBERBLOCK_SHIFT(ashift))
+
+#define	LABEL_SIZE 262144
+
+static void
+zhack_repair_one_label_cksum(const int fd,
+    vdev_label_t * const vl,
+    const uint64_t label_offset,
+    const int l,
+    uint32_t * const labels_repaired)
 {
-	zio_checksum_info_t *ci = &zio_checksum_table[ZIO_CHECKSUM_LABEL];
 	const char *cfg_keys[] = { ZPOOL_CONFIG_VERSION,
 	    ZPOOL_CONFIG_POOL_STATE, ZPOOL_CONFIG_GUID };
-	boolean_t labels_repaired[VDEV_LABELS] = {0};
-	boolean_t repaired = B_FALSE;
-	vdev_label_t labels[VDEV_LABELS] = {{{0}}};
-	struct stat st;
-	int fd;
+	uberblock_t *ub;
+	uint64_t ashift;
+	int byteswap;
+	zio_eck_t *ub_eck;
+	zio_eck_t *vdev_eck;
+	nvlist_t *cfg;
+	nvlist_t *vdev_tree_cfg;
+	uint64_t txg;
+	char *buf;
+	size_t buflen;
+	void *ub_data;
+	void *vdev_data;
+	ssize_t err;
+
+	ub = (uberblock_t *)vl->vl_uberblock;
+
+	err = pread64(fd, vl, sizeof (vdev_label_t), label_offset);
+	if (err == -1) {
+		(void) fprintf(stderr,
+		    "error: cannot read label %d: %s\n",
+		    l, strerror(errno));
+		return;
+	} else if (err != sizeof (vdev_label_t)) {
+		(void) fprintf(stderr,
+		    "error: bad label %d read size \n", l);
+		return;
+	}
 
-	abd_init();
+	err = nvlist_unpack(vl->vl_vdev_phys.vp_nvlist,
+	    VDEV_PHYS_SIZE - sizeof (zio_eck_t), &cfg, 0);
+	if (err) {
+		(void) fprintf(stderr,
+		    "error: cannot unpack nvlist label %d\n", l);
+		return;
+	}
 
-	argc -= 1;
-	argv += 1;
+	if (ub->ub_txg != 0) {
+		(void) fprintf(stderr,
+		    "error: label %d: UB TXG of 0 expected, but got %"
+		    PRIu64 "\n",
+		    l, ub->ub_txg);
+		return;
+	}
```

Comment:
It would be helpful if you could add some comments explaining the rational for what's being checked. For example, the `txg == 0` check here is used as an indicator that the vdev was cleanly initialized and the first uberblock index is expected to contain the latest valid uberblock.  See `vdev_label_init()` which is called with the `VDEV_LABEL_REMOVE` reason on detach.

Reply:
I do believe I misunderstood what that check was for when I first wrote it, to be honest. I assumed txg should be zero when the device was attached, but this is actually done every time the device is exported?

Also, perhaps it would be preferable to make a more descriptive error message that informs both the user and the programmer what the check is for?

Reply:
> I assumed txg should be zero when the device was detached

This should be true, but only because when a device is detached with always reinitialize that label but also make a copy of the latest block pointer as you saw.  During an export the label is also updated with it will be written with the latest txg.